### PR TITLE
score and levels

### PR DESCRIPTION
**Pull request by willharrison**

### CODE REVIEW

File: js/app/config.js [hunk 1]
```diff
@@ -56,7 +59,10 @@ define(function(){
                 decayRate : 0.95,
                 minThreshold : 60
             }
-        }
+        },
+        points : [ 1.5, 2, 2.5, 3 ],
+        checkpoints : [ 100, 500, 5000 ],
+        speeds : [ 4, 3, 2, 1 ]
```
Comment (ALSchwalm) [inline js/app/config.js]: Is it possible to play the game at all at this top speed?

File: js/app/block.js [hunk 2]
```diff
@@ -239,11 +248,18 @@ define(["app/config", "app/grid"], function(config, grid){
             var totalCleared = eraseBlocks(this) + grid.cleanup();
             this.displayClearedCount(totalCleared);
             this.game.add.audio('destroy').play();
+            this.game.scoreboard.update(totalCleared);
+            this.game.scoreboard.currentLevel = this.game.scoreboard.calcLevel();
```
Comment (ALSchwalm) [inline js/app/block.js]: Can this not just happen inside of the scoreboard.update?

File: js/app/config.js (hunk 1 above)
Comment (willharrison) [inline js/app/config.js]: haven't made it past level 3, so I wouldn't know.  Also, the checkpoints, points, and speeds aren't really important.  I figure that we can tweak them.

File: js/app/generator.js [hunk 3]
```diff
@@ -174,12 +178,24 @@ function(config, Quad, grid){
...
+    /**
      * Draw the graphics showing the time remaining before the next drop
      */
     Generator.prototype.drawTimerGraphics = function() {
+        var speed = config.speeds[this.level];
+
         this.timerGraphic.clear();
         var percentElapsed =
-                this.dropTimer.ms/(config.generator.defaultWait*1000)
+                this.dropTimer.ms/(speed * 1000)
```
Comment (ALSchwalm) [inline js/app/generator.js]: Please use the Phaser.Time.SECOND constant here as well (I should have originally)

File: js/app/grid.js [hunk 4]
```diff
@@ -343,10 +343,10 @@ function(config, Phaser, music){
         for (var i=0; i < config.grid.numCells; ++i) {
             for (var j=0; j < config.grid.numCells; ++j) {
                 if (this.at(i, j) &&
-                    !this.at(i+1, j) && !this.at(i, j+1) &&
-                    !this.at(i-1, j) && !this.at(i, j-1)) {
-                    this.at(i, j).destroy();
-                    totalCleared += 1;
+                !this.at(i+1, j) && !this.at(i, j+1) &&
```
Comment (ALSchwalm) [inline js/app/grid.js]: I'm not sure about this indentation

File: js/app/score.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        };
+        this.board = game.add.text(game.world.centerX + 150, 27, text, style);
+        this.grid = grid;
+        this.generator = generator;
+    }
+
+    /**
+     * Get current score.
+     */
+    Score.prototype.getCurrent = function() {
```
Comment (ALSchwalm) [inline js/app/score.js]: Why do we need a getter here?

File: js/app/score.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+     * Calculate what level we are on now.
+     */
+    Score.prototype.calcLevel = function() {
+        for (var i = config.checkpoints.length - 1; i >= 0; i--) {
+            if (this.current >= config.checkpoints[i]) {
+                if (this.level < i + 1) {
+                    this.level = i + 1;
+                    this.generator.updateCurrentQuadLevel(this.level);
+                    this.update(0, this.level);
+                    console.log('hit');
```
Comment (ALSchwalm) [inline js/app/score.js]: No console.log in master

File: js/app/state/create.js [hunk 7]
```diff
@@ -18,6 +18,10 @@ function(grid, generator, background, music){
         background.start(game);
         grid.display(game);
         generator.start(game);
+        score.init(game, grid, generator);
+        
+        game.scoreboard = score;
+        game.generator = generator;
```
Comment (ALSchwalm) [inline js/app/state/create.js]: We should not be attaching anything to the game object. If a module needs access to the scoreboard or generator, it should depend on them explicitly, either by requiring them, or accepting them at construction/initialization.

Comment (willharrison) [inline js/app/state/create.js]: that sucks.  you going to merge it?

File: js/app/grid.js [hunk 8]
```diff
@@ -354,6 +354,20 @@ function(config, Phaser, music){
     }
 
     /**
+     * Clear all blocks on the grid.
+     */
+    Grid.prototype.clearAll = function() {
```
Comment (ALSchwalm) [inline js/app/grid.js]: Perhaps this should be called 'destroyAll', as it does not invoke clear on the blocks.

Comment (willharrison) [inline js/app/grid.js]: perhaps.  I think it'll be fine, though.

Comment (ALSchwalm) [inline js/app/grid.js]: This also needs to recenter the unbreakable blocks

**Comment by willharrison** [comment]
I cleaned up some of the stuff.  I will work on reseting unbreakable blocks tomorrow night.

**Comment by ALSchwalm** [comment]
Merging this, you can make the resetting blocks a separate PR

**Pull request merged**